utils/options/options.go

```diff
@@ -0,0 +1,22 @@
+// Copyright 2023 LiveKit, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package options
+
+func Apply[T any, F ~func(T)](o T, opts []F) T {
```

Comment:
Sorry Paul, I already asked you this once, but I forgot the answer :-(. What does the `~` mean there?

Reply:
`~` means the type constraint permits any type with the same underlying type. so for example
```go
type Foo func()

var f0 = func() {}
var f1 Foo = func() {}

func a[T ~func()](f T) {}
func b[T func()](f T)  {}

a(f0) // ok
b(f0) // ok
a(f1) // ok
b(f1) // not ok
```

Reply:
ah, got it. Thank you Paul. Let me commit this to my poor memory and hope it survives till the next time I see this pattern :-)

Reply:
this is needed here because we have functional options like
```go
type FooOption = func(x *FooOptions)

// several implementations of FooOption

func SomeFoo(opts ...FooOption) {
    _ := options.Apply(&FooOptions{}, opts)
    // without ~ this wouldn't work because []FooOption is not []func(*FooOptions)
}
```